rust-runtime/smithy-json/src/serialize.rs

```diff
@@ -0,0 +1,241 @@
+        self
+    }
+
+    /// Writes a number `value` (already converted to string) with the given `key`.
+    pub fn number(&mut self, key: &str, value: &str) -> &mut Self {
+        self.key(key);
+        self.json.push_str(value);
+        self
+    }
+
+    /// Starts an array with the given `key`.
+    pub fn start_array(&mut self, key: &str) -> JsonArrayWriter {
+        self.key(key);
+        JsonArrayWriter::new(&mut self.json)
+    }
+
+    /// Starts an object with the given `key`.
+    pub fn start_object(&mut self, key: &str) -> JsonObjectWriter {
+        self.key(key);
+        JsonObjectWriter::new(&mut self.json)
+    }
+
+    /// Finishes the object.
+    pub fn finish(self) {
+        self.json.push('}');
+    }
+
+    fn key(&mut self, key: &str) {
+        if self.started {
+            self.json.push(',');
+        }
+        self.started = true;
+
+        self.json.push('"');
+        self.json.push_str(&escape_string(key));
+        self.json.push_str("\":");
+    }
+}
+
+pub struct JsonArrayWriter<'a> {
+    json: &'a mut String,
+    started: bool,
+}
+
+impl<'a> JsonArrayWriter<'a> {
+    pub fn new(output: &'a mut String) -> Self {
+        output.push('[');
+        Self {
+            json: output,
+            started: false,
+        }
+    }
+
+    /// Writes a null value to the array.
+    pub fn null(&mut self) -> &mut Self {
+        self.comma_delimit();
+        self.json.push_str("null");
+        self
+    }
+
+    /// Writes the boolean `value` to the array.
+    pub fn boolean(&mut self, value: bool) -> &mut Self {
+        self.comma_delimit();
+        self.json.push_str(match value {
+            true => "true",
+            _ => "false",
+        });
+        self
+    }
+
+    /// Writes a string to the array.
+    pub fn string(&mut self, value: &str) -> &mut Self {
+        self.comma_delimit();
+        append_string(&mut self.json, value);
+        self
+    }
+
+    /// Writes a number `value` (already converted to string).
+    pub fn number(&mut self, value: &str) -> &mut Self {
+        self.comma_delimit();
+        self.json.push_str(value);
+        self
+    }
+
+    /// Starts a nested array inside of the array.
+    pub fn start_array(&mut self) -> JsonArrayWriter {
+        self.comma_delimit();
+        JsonArrayWriter::new(&mut self.json)
+    }
+
+    /// Starts a nested object inside of the array.
+    pub fn start_object(&mut self) -> JsonObjectWriter {
+        self.comma_delimit();
+        JsonObjectWriter::new(&mut self.json)
+    }
+
+    /// Finishes the array.
+    pub fn finish(self) {
+        self.json.push(']');
+    }
+
+    fn comma_delimit(&mut self) {
+        if self.started {
+            self.json.push(',');
+        }
+        self.started = true;
+    }
+}
+
+fn append_string(json: &mut String, value: &str) {
+    json.push('"');
+    json.push_str(&escape_string(value));
+    json.push('"');
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
can you add test coverage for `null`/`boolean`? All the other methods are covered